File: algo/src/BOJ/Gold/BOJ13549.java
package BOJ.Gold;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.PriorityQueue;
import java.util.StringTokenizer;

public class BOJ13549 {

    private static int[] arr;
    private static int[] visited;
    private static int n;
    private static int k;

    private static class info implements Comparable<info>{
        int pos;
        int turn;

        info(int pos,int turn){
            this.pos = pos;
            this.turn = turn;
        }

        @Override
        public int compareTo(info o) {
            return this.turn-o.turn;
        }
    }
    public static void main(String[] args) throws IOException {
        BufferedReader br =new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        n = Integer.parseInt(st.nextToken());
        k = Integer.parseInt(st.nextToken());
        arr = new int[100001];
        visited = new int[100001];
        Arrays.fill(visited,987654321);
        bfs();
        System.out.println(visited[k]);
    }

    private static void bfs() {
        PriorityQueue<info> que = new PriorityQueue<>();
        que.add(new info(n,0));

        while(!que.isEmpty()){
            info z = que.poll();

            if(visited[z.pos]<z.turn)continue;
            visited[z.pos]=z.turn;

            if(z.pos*2<=100000){
                que.add(new info(z.pos*2,z.turn));

            }
            if(z.pos+1<=100000){
                que.add(new info(z.pos+1,z.turn+1));
            }
            if(z.pos-1>=0){
                que.add(new info(z.pos-1,z.turn+1));
            }

        }
    }
}

File: PS/algo/src/BOJ/Gold/BOJ13549.kt
package BOJ.gold

import java.util.PriorityQueue

private val size = 100_001
fun main() {
    val br = System.`in`.bufferedReader()
    val (n,k) = br.readLine().split(" ").map { it.toInt() }
    val visited =  BooleanArray(size){false}
    val que = PriorityQueue<info>()

    que.add(info(n,0))
    visited[n]=true

    while(que.isNotEmpty()){
        val z:info = que.poll()
        visited[z.pos]=true

        if(z.pos==k) return println(z.cost)

        if(z.pos * 2<size && !visited[z.pos * 2] ) que.add(info(z.pos * 2,z.cost))
        if(z.pos-1>=0 && !visited[z.pos-1]) que.add(info(z.pos-1,z.cost+1))
        if(z.pos+1<size && !visited[z.pos+1]) que.add(info(z.pos+1,z.cost+1))

    }

}

private data class info(val pos:Int,val cost:Int): Comparable<info>{
    override fun compareTo(other: info): Int {
        return cost-other.cost
    }
}